### PR TITLE
Maintenance: Prepare migration of NSKeyedArchiver methods and move to block-based animation

### PR DESCRIPTION
**Pull request by wutschel**
## Description

Closes https://github.com/xbmc/Official-Kodi-Remote-iOS/issues/599.

This PR has four main sections:
1. Prepare migration of `NSKeyedArchiver` methods
The method `unarchiveObjectWithFile` and `archiveRootObject:toFile:` are deprecated since iOS 11. This PR refactors reading/writing persistent data and already prepares the migration to the newly introduced `unarchivedObjectOfClasses` and `archivedDataWithRootObject:requiringSecureCoding:error:`. 
2. Move duplicate animation related code to Utilities
This allows to remove code duplication and unneeded code, as well as prepares for the next third step.
3. Migrate to block-based animation
All parameters and options were taken over into the block-based animation.
4. Optimizations and bugfixes around animation in `NowPlaying`
The code was refactored to remove unneeded variables. In addition, the animation for the transition between playlist and NowPlaying view was corrected (now properly sliding out and in). As a further optimization the "demo" animation for the playlist/NowPlaying toggle button is starting earlier. This always bugged me.

## Summary for release notes

Maintenance: Prepare migration of NSKeyedArchiver methods
Maintenance: Migrate to block-based animation
Bugfix: Fix flip in/out animation for playlist/NowPlaying view

### CODE REVIEW

**Comment by wutschel** [comment]
@kambala-decapitator, I am keeping this as is for the review. There are multiple small commits which should make this PR easier to review. One this is done I propose to squash all commits of section 2 and section 3 into a single commit each.

File: XBMC Remote/ViewControllerIPad.m [hunk 1]
```diff
@@ -609,12 +609,14 @@ - (void)handleTcpJSONRPCChangeServerStatus:(NSNotification*)sender {
...
-    [UIView setAnimationDuration:0.2];
-    self.nowPlayingController.songDetailsView.alpha = 0.0;
-    self.nowPlayingController.itemDescription.scrollsToTop = NO;
-    [UIView commitAnimations];
+    [UIView animateWithDuration:0.2
+                          delay:0.0
+                        options:UIViewAnimationOptionCurveEaseInOut
+                     animations:^{
+        self.nowPlayingController.songDetailsView.alpha = 0.0;
+        self.nowPlayingController.itemDescription.scrollsToTop = NO;
```
Comment (kambala-decapitator) [inline XBMC Remote/ViewControllerIPad.m:617]: this property is not animatable

File: XBMC Remote/RemoteController.m [hunk 2]
```diff
@@ -402,21 +402,25 @@ - (void)touchesCancelled:(NSSet*)touches withEvent:(UIEvent*)event {
...
-    [UIView setAnimationCurve:UIViewAnimationCurveEaseInOut];
-	[UIView setAnimationDuration:0.1];
-    subsInfoLabel.alpha = 0;
-    [UIView commitAnimations];
+    // first fadeout
+    [UIView animateWithDuration:0.1
+                          delay:0.0
+                        options:UIViewAnimationOptionCurveEaseInOut
+                     animations:^{
+        subsInfoLabel.alpha = 0;
```
Comment (kambala-decapitator) [inline XBMC Remote/RemoteController.m:410]: you already have utility method to animate alpha, why not using it? there're also other places like this.

File: XBMC Remote/NowPlaying.m [hunk 3]
```diff
@@ -1438,18 +1440,20 @@ - (void)flipAnimButton:(UIButton*)button demo:(BOOL)demo {
                              [button setImage:image forState:UIControlStateHighlighted];
                              [button setImage:image forState:UIControlStateSelected];
                          }
-                         [UIView setAnimationCurve:UIViewAnimationCurveEaseIn];
                          [UIView setAnimationTransition:anim forView:button cache:YES];
```
Comment (kambala-decapitator) [inline XBMC Remote/NowPlaying.m]: is this line removed in next commit?

File: XBMC Remote/NowPlaying.m [hunk 4]
```diff
@@ -1438,18 +1440,20 @@ - (void)flipAnimButton:(UIButton*)button demo:(BOOL)demo {
...
                      } 
                      completion:^(BOOL finished) {
-                         [UIView beginAnimations:nil context:nil];
-                         button.hidden = NO;
-                         [UIView setAnimationDuration:0.5];
-                         [UIView setAnimationDelegate:self];
```
Comment (kambala-decapitator) [inline XBMC Remote/NowPlaying.m:1485]: I don't see delegate method being removed / its logic being ported. Or is there no such method?

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]
I'd squash most of the animation commits into a single one: now you have utility methods for label / table / button / view, but all of these elements are views, don't see a reason to have dedicated methods that essentially do the same thing.

also it seems that in some places indentation broke.

File: XBMC Remote/ViewControllerIPad.m (hunk 1 above)
Comment (wutschel) [reply XBMC Remote/ViewControllerIPad.m:617]: Does it harm to keep `self.nowPlayingController.itemDescription.scrollsToTop = NO;` in the animation block? Or does it need to be moved out? Before this change it was also in the animated block.

File: XBMC Remote/NowPlaying.m (hunk 4 above)
Comment (wutschel) [reply XBMC Remote/NowPlaying.m:1485]: There is nothing like this. I guess this was just superfluous.

File: XBMC Remote/Utilities.m [hunk 5]
```diff
@@ -977,4 +977,41 @@ + (void)enableDefaultController:(id<UITableViewDelegate>)viewController tableVie
...
+        return nil;
+    } else {
+        return unarchived;
+    }
+}
+
++ (void)archivePath:(NSString*)path file:(NSString*)filename data:(id)data {
+    NSString *filePath = [path stringByAppendingPathComponent:filename];
+    NSError *error;
+    NSData *archiveData;
```
[the remark this answers is not in the document]
Comment (wutschel) [reply XBMC Remote/Utilities.m:1005]: Done

File: XBMC Remote/Utilities.m [hunk 6]
```diff
@@ -977,4 +977,41 @@ + (void)enableDefaultController:(id<UITableViewDelegate>)viewController tableVie
...
+                                                            error:&error];
+    } else {
+        unarchived = [NSKeyedUnarchiver unarchiveObjectWithFile:filePath];
+    }
+    
+    if (error) {
+        return nil;
+    } else {
+        return unarchived;
+    }
```
[the remark this answers is not in the document]
Comment (wutschel) [reply XBMC Remote/Utilities.m:999]: Done

File: XBMC Remote/Utilities.m [hunk 7]
```diff
@@ -977,4 +977,41 @@ + (void)enableDefaultController:(id<UITableViewDelegate>)viewController tableVie
...
 
++ (id)unarchivePath:(NSString*)path file:(NSString*)filename {
+    NSString *filePath = [path stringByAppendingPathComponent:filename];
+    NSData *data = [[NSFileManager defaultManager] contentsAtPath:filePath];
+    NSArray *objectClasses = @[[NSObject class]];
+    NSError *error;
+    id unarchived;
+    
+    if (@available(iOS 11.0, *)) {
+        unarchived = [NSKeyedUnarchiver unarchivedObjectOfClasses:[NSSet setWithArray:objectClasses]
```
[the remark this answers is not in the document]
Comment (wutschel) [reply XBMC Remote/Utilities.m:988]: True, changed it.

File: XBMC Remote/AppDelegate.m [hunk 8]
```diff
@@ -384,27 +384,30 @@ - (NSDictionary*)sortmethod:(NSString*)method order:(NSString*)order ignoreartic
...
-        NSString *fullNamespace = @"LibraryCache";
+        
+        NSFileManager *fileManager = [NSFileManager defaultManager];
         paths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
-        self.libraryCachePath = [paths[0] stringByAppendingPathComponent:fullNamespace];
-        if (![fileManager1 fileExistsAtPath:self.libraryCachePath]) {
-            [fileManager1 createDirectoryAtPath:self.libraryCachePath withIntermediateDirectories:YES attributes:nil error:NULL];
+        self.libraryCachePath = [paths[0] stringByAppendingPathComponent:@"LibraryCache"];
+        if (![fileManager fileExistsAtPath:self.libraryCachePath]) {
+            [fileManager createDirectoryAtPath:self.libraryCachePath
```
[the remark this answers is not in the document]
Comment (wutschel) [reply XBMC Remote/AppDelegate.m:400]: Ok, removed the check -- also for the 2nd case.

File: XBMC Remote/AppDelegate.m [hunk 9]
```diff
@@ -384,27 +384,30 @@ - (NSDictionary*)sortmethod:(NSString*)method order:(NSString*)order ignoreartic
 - (id)init {
 	if ((self = [super init])) {
         NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
-        NSString *documentsDirectory = paths[0];
-        self.dataFilePath = [documentsDirectory stringByAppendingPathComponent:@"serverList_saved.dat"];
-        NSFileManager *fileManager1 = [NSFileManager defaultManager];
-        if ([fileManager1 fileExistsAtPath:self.dataFilePath]) {
-            NSMutableArray *tempArray;
-            tempArray = [NSKeyedUnarchiver unarchiveObjectWithFile:self.dataFilePath];
+        self.dataFilePath = paths[0];
```
[the remark this answers is not in the document]
Comment (wutschel) [reply XBMC Remote/AppDelegate.m:387]: Agreed, done.

File: XBMC Remote/AppDelegate.m [hunk 10]
```diff
@@ -384,27 +384,30 @@ - (NSDictionary*)sortmethod:(NSString*)method order:(NSString*)order ignoreartic
...
+            [fileManager createDirectoryAtPath:self.libraryCachePath
+                   withIntermediateDirectories:YES
+                                    attributes:nil
+                                         error:NULL];
         }
         self.epgCachePath = [paths[0] stringByAppendingPathComponent:@"EPGDataCache"];
-//        [[NSFileManager defaultManager] removeItemAtPath:self.epgCachePath error:nil];
-        if (![fileManager1 fileExistsAtPath:self.epgCachePath]) {
-            [fileManager1 createDirectoryAtPath:self.epgCachePath withIntermediateDirectories:YES attributes:nil error:NULL];
+        if (![fileManager fileExistsAtPath:self.epgCachePath]) {
```
[the remark this answers is not in the document]
Comment (wutschel) [reply XBMC Remote/AppDelegate.m:406]: But it also does not harm to keep this. If you don't mind, let us not change.

File: XBMC Remote/RemoteController.m (hunk 2 above)
Comment (wutschel) [reply XBMC Remote/RemoteController.m:410]: > I'd squash most of the animation commits into a single one

Yes, I wanted to squash 
1. [Use the block-based animation API] up to [Use the block-based animation API (7)]2. 
2. [Move AnimTable] up to [Remove alphaButton]

> now you have utility methods for label / table / button / view, but all of these elements are views, don't see a reason to have dedicated methods that essentially do the same thing.

Hmm, I didn't look into this yet. Let me see.

> also it seems that in some places indentation broke.

Well, especially `StackScrollViewController` has completely messed up indentations. Shall I just auto-indent this in a commit at the end?

**Comment by wutschel** [comment]
> II'd squash most of the animation commits into a single one

Yes, I wanted to squash
1. [Use the block-based animation API] up to [Use the block-based animation API (7)]2.
2. [Move AnimTable] up to [Remove alphaButton]3. 

> now you have utility methods for label / table / button / view, but all of these elements are views, don't see a reason to have dedicated methods that essentially do the same thing.

Hmm, I didn't look into this yet. Let me see.

> also it seems that in some places indentation broke.

Well, especially `StackScrollViewController` has completely messed up indentations. Shall I just auto-indent this file in a commit at the end?

File: XBMC Remote/RemoteController.m (hunk 2 above)
Comment (wutschel) [reply XBMC Remote/RemoteController.m:410]: Is `UIViewAnimationOptionCurveEaseInOut` the default option?

Comment (wutschel) [reply XBMC Remote/RemoteController.m]: It is `UIViewAnimationOptionCurveEaseInOut = 0 << 16`.

Nevertheless explicitly added `UIViewAnimationOptionCurveEaseInOut` as option the the helper methods.

File: XBMC Remote/ViewControllerIPad.m (hunk 1 above)
Comment (wutschel) [reply XBMC Remote/ViewControllerIPad.m]: Done

File: XBMC Remote/AppDelegate.m (hunk 10 above)
[the remark this answers is not in the document]
Comment (kambala-decapitator) [reply XBMC Remote/AppDelegate.m]: sure, I don't insist

File: XBMC Remote/ViewControllerIPad.m (hunk 1 above)
Comment (kambala-decapitator) [reply XBMC Remote/ViewControllerIPad.m]: no harm here, I just prefer that animation block contains only animations to make things clearer

File: XBMC Remote/RemoteController.m (hunk 2 above)
Comment (kambala-decapitator) [reply XBMC Remote/RemoteController.m]: I was thinking about second method that accepts the curve parameter, will check your solution

File: XBMC Remote/NowPlaying.m (hunk 3 above)
Comment (kambala-decapitator) [reply XBMC Remote/NowPlaying.m]: thanks, was difficult to notice in that huge commit

**Comment by wutschel** [comment]
Anything more I should work on? Or are we good and I can start to squash and rebase?

File: XBMC Remote/Utilities.m [hunk 11]
```diff
@@ -1031,25 +1023,12 @@ + (void)AnimView:(UIView*)view AnimDuration:(NSTimeInterval)seconds Alpha:(CGFlo
...
 }
 
 + (void)AnimButton:(UIButton*)button AnimDuration:(NSTimeInterval)seconds hidden:(BOOL)hiddenValue XPos:(int)X {
-    [UIView animateWithDuration:seconds
-                     animations:^{
-        CGRect frame;
-        frame = button.frame;
-        frame.origin.x = X;
-        button.frame = frame;
-        button.hidden = hiddenValue;
```
Comment (kambala-decapitator) [inline XBMC Remote/Utilities.m]: - `hidden` isn't an animatable property
- you replaced `hidden` with `alpha` (which is animatable), is it intended?

File: XBMC Remote/Utilities.m [hunk 12]
```diff
@@ -1060,13 +1039,7 @@ + (void)alphaView:(UIView*)view AnimDuration:(NSTimeInterval)seconds Alpha:(CGFl
 }
 
 + (void)alphaImage:(UIImageView*)image AnimDuration:(NSTimeInterval)seconds Alpha:(CGFloat)alphavalue {
-    [UIView animateWithDuration:seconds
-                     animations:^{
-        image.alpha = alphavalue;
-        if (alphavalue) {
-            image.hidden = NO;
```
Comment (kambala-decapitator) [inline XBMC Remote/Utilities.m]: also hidden/alpha point

File: XBMC Remote/RemoteController.m (hunk 2 above)
Comment (kambala-decapitator) [reply XBMC Remote/RemoteController.m]: yeah, always this param is used, no need for a second method

**Comment by kambala-decapitator** [comment]
squash: sounds good

indent: yes, let's do that

File: XBMC Remote/Utilities.m (hunk 11 above)
Comment (wutschel) [reply XBMC Remote/Utilities.m]: Yes, this was intended.

File: XBMC Remote/Utilities.m (hunk 12 above)
Comment (wutschel) [reply XBMC Remote/Utilities.m]: Yes, I was taking care if this. `alphaImage:` was only called in few places. The view handed over was either `fanartView` which never is hidden, or `jewelView` for which I added an explicit unhide before calling the unified method `alphaView:`. So, the `hidden = NO` is not required inside the helper method anymore.

**Comment by wutschel** [comment]
Squashed and rebased. I skipped the auto-indent as I forgot I already have some work prepared in the background which conflicts with this. I will do this next time I am touching this file as part of a PR.

**Review by kambala-decapitator — APPROVED** [review]
just to double check: this is ready to be merged, isn't it?

**Comment by wutschel** [comment]
> just to double check: this is ready to be merged, isn't it?

Yes, it is.

**Comment by kambala-decapitator** [comment]
@wutschel have you seen this warning in the console?

> NSSecureCoding allowed classes list contains [NSObject class], which bypasses security by allowing any Objective-C class to be implicitly decoded. Consider reducing the scope of allowed classes during decoding by listing only the classes you expect to decode, or a more specific base class than NSObject.

**Comment by wutschel** [comment]
> I guess it can be ignored since secure coding isn't used?

I have seen this yesterday for the first time. Not sure why it was not coming up before. Maybe it only showed after I removed the App.

I do not know what this can cause. I'd rather take a deeper look. In worst case I simply provide a PR which comments the new code for 1.10 to give time to deal this properly.

**Comment by kambala-decapitator** [comment]
I think there's no issue, but please investigate shortly to be 100% sure

**Pull request merged**